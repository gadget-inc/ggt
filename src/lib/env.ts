import type { Paths } from "env-paths";
import envPaths from "env-paths";

/**
 * Captures the name and nature of the environment
 */
export class Env {
  static get value(): string {
    return process.env["GGT_ENV"] || "production";
  }

  static get paths(): Paths {
    return envPaths("gadget", { suffix: this.productionLike ? "" : this.value });
  }

  static get developmentLike(): boolean {
    return this.value.startsWith("development");
  }

  static get testLike(): boolean {
    return this.value.startsWith("test");
  }

  static get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  }

  static get productionLike(): boolean {
    return this.value.startsWith("production");
  }
}
