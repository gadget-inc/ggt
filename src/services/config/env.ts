import process from "node:process";

/**
 * Captures the name and nature of the environment
 */
export const env = {
  get value(): string {
    return process.env["GGT_ENV"] || "production";
  },

  get productionLike(): boolean {
    return !this.developmentOrTestLike;
  },

  get developmentLike(): boolean {
    return this.value.startsWith("development");
  },

  get testLike(): boolean {
    return this.value.startsWith("test");
  },

  get developmentOrTestLike(): boolean {
    return this.developmentLike || this.testLike;
  },
};
