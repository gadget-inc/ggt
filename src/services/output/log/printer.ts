import type { SetReturnType } from "type-fest";
import { config } from "../../config/config.js";
import { isString } from "../../util/is.js";
import { sprint, sprintTable, sprintln, type Sprint, type SprintOptions, type SprintTableOptions } from "../sprint.js";
import { stdout } from "../stream.js";
import { formatters } from "./format/format.js";
import { Level } from "./level.js";

export type Print = {
  (options: SprintOptions): SetReturnType<Sprint, void>;
  (template: TemplateStringsArray | string, ...values: unknown[]): void;
};

export type Println = {
  (options: SprintOptions): SetReturnType<Sprint, void>;
  (template?: TemplateStringsArray | string, ...values: unknown[]): void;
};

export type Printer = {
  /**
   * Prints a message to stdout.
   *
   * @example
   * logger.print("Hello, world!");
   * // Hello, world!
   */
  print: Print;

  /**
   * Prints a message to stdout followed by a newline.
   *
   * @example
   * logger.println("Hello, world!");
   * // Hello, world!\n
   */
  println: Println;

  /**
   * Prints a table to stdout.
   */
  printTable: (options: SprintTableOptions) => void;

  /**
   * Creates a buffer that can be used to batch multiple print
   * statements together and flush them all at once.
   */
  buffer(): BufferedPrinter;
};

export type BufferedPrinter = Omit<Printer, "buffer"> & {
  /**
   * Flushes the buffer to stdout.
   */
  flush(): void;

  /**
   * Returns the buffer as a string.
   */
  toString(): string;
};

export const createPrinter = ({ name }: { name: string }): Printer => {
  const printMsg = (msg: string): void => {
    if (config.logLevel < Level.PRINT || config.logFormat === "json") {
      msg = formatters[config.logFormat](Level.PRINT, name, msg, {});
    }
    stdout.write(msg);
  };

  return {
    print: ((...args) => {
      // @ts-expect-error - TS doesn't like the overloads here
      const result = sprint(...args);
      if (isString(result)) {
        printMsg(result);
        return;
      }

      // @ts-expect-error - or here
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      return (...args) => printMsg(result(...args));
    }) as Print,

    println: ((...args) => {
      // @ts-expect-error - TS doesn't like the overloads here
      const result = sprintln(...args);
      if (isString(result)) {
        printMsg(result);
        return;
      }

      // @ts-expect-error - or here
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      return (...args) => printMsg(result(...args));
    }) as Println,

    printTable(opts) {
      if (config.logFormat === "json") {
        stdout.write(
          formatters.json(Level.PRINT, name, opts.message || opts.boxen?.title || "table", {
            headers: opts.headers,
            rows: opts.rows,
            footer: opts.footer,
          }),
        );
        return;
      }

      this.println(sprintTable(opts));
    },

    buffer: () => {
      let buf: string[] = [];

      return {
        print: ((...args) => {
          // @ts-expect-error - TS doesn't like the overloads here
          buf.push(sprint(...args));
        }) as Print,
        println: ((...args) => {
          // @ts-expect-error - or here
          buf.push(sprintln(...args));
        }) as Println,
        printTable: (options) => {
          buf.push(sprintTable(options));
        },
        toString: () => {
          return buf.join("");
        },
        flush() {
          for (const msg of buf) {
            stdout.write(msg);
          }
          buf = [];
        },
      };
    },
  };
};
