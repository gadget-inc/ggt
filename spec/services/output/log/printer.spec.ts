import chalk from "chalk";
import { describe, it, vi } from "vitest";
import { createPrinter } from "../../../../src/services/output/log/printer.js";
import { withEnv } from "../../../__support__/env.js";
import { expectStdout } from "../../../__support__/stdout.js";

describe("printer", () => {
  const printer = createPrinter({ name: "printer" });

  for (const method of ["print", "println", "printlns"] as const) {
    describe(method, () => {
      it("writes to stdout", () => {
        printer[method]("Hello, world!");
        expectStdout().toMatchSnapshot();
      });

      it("strips indentation from multiline strings", () => {
        printer[method](`
            Hello, world!
            I am a multiline string.
        `);
        expectStdout().toMatchSnapshot();
      });

      it("supports tagged template literals", () => {
        printer[method]`
            Hello, world!
            I am a multiline string.
        `;
        expectStdout().toMatchSnapshot();
      });

      it("supports chalk colors within template literals", () => {
        try {
          chalk.level = 3;
          printer[method]`
            {bold Hello, world!}
            {underline I am a multiline string.}
          `;
          expectStdout().toMatchSnapshot();
        } finally {
          chalk.level = 0;
        }
      });

      it("prints like a structured logger when GGT_LOG_LEVEL is set", () => {
        try {
          vi.useFakeTimers();
          vi.setSystemTime(0);

          withEnv({ GGT_LOG_LEVEL: "info" }, () => {
            printer[method]("Hello, world!");
            expectStdout().toMatchSnapshot();
          });
        } finally {
          vi.useRealTimers();
        }
      });

      it("prints json when GGT_LOG_FORMAT=json", () => {
        withEnv({ GGT_LOG_FORMAT: "json" }, () => {
          printer[method]("Hello, world!");
          expectStdout().toMatchSnapshot();
        });
      });
    });
  }

  describe("printTable", () => {
    it("writes a table to stdout", () => {
      printer.printTable({
        message: "Table Title",
        headers: ["Column 1", "Column 2"],
        rows: [
          ["Row 1, Column 1", "Row 1, Column 2"],
          ["Row 2, Column 1", "Row 2, Column 2"],
        ],
        footer: "Table Footer",
      });
      expectStdout().toMatchSnapshot();
    });

    it("writes a table with without a footer to stdout", () => {
      printer.printTable({
        message: "Table Title",
        headers: ["Column 1", "Column 2"],
        rows: [
          ["Row 1, Column 1", "Row 1, Column 2"],
          ["Row 2, Column 1", "Row 2, Column 2"],
        ],
      });
      expectStdout().toMatchSnapshot();
    });

    it("writes the table as json when GGT_LOG_FORMAT=json", () => {
      withEnv({ GGT_LOG_FORMAT: "json" }, () => {
        printer.printTable({
          message: "Table Title",
          headers: ["Column 1", "Column 2"],
          rows: [
            ["Row 1, Column 1", "Row 1, Column 2"],
            ["Row 2, Column 1", "Row 2, Column 2"],
          ],
          footer: "Table Footer",
        });
        expectStdout().toMatchSnapshot();
      });
    });

    for (const borders of ["none", "thin", "thick"] as const) {
      it(`writes a table with ${borders} borders to stdout`, () => {
        printer.printTable({
          borders: borders,
          message: "Table Title",
          headers: ["Column 1", "Column 2"],
          rows: [
            ["Row 1, Column 1", "Row 1, Column 2"],
            ["Row 2, Column 1", "Row 2, Column 2"],
          ],
          footer: "Table Footer",
        });
        expectStdout().toMatchSnapshot();
      });
    }
  });
});
