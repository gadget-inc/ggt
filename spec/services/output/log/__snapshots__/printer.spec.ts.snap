// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`printer > print > prints json when GGT_LOG_FORMAT=json 1`] = `
"{\\"level\\":6,\\"name\\":\\"printer\\",\\"msg\\":\\"Hello, world!\\",\\"fields\\":{}}
"
`;

exports[`printer > print > prints like a structured logger when GGT_LOG_LEVEL is set 1`] = `
"12:00:00 PRINT printer: Hello, world!
"
`;

exports[`printer > print > strips indentation from multiline strings 1`] = `
"Hello, world!
I am a multiline string."
`;

exports[`printer > print > supports chalk colors within template literals 1`] = `
"[1mHello, world![22m
[4mI am a multiline string.[24m"
`;

exports[`printer > print > supports margin top 1`] = `
"
Hello, world!"
`;

exports[`printer > print > supports tagged template literals 1`] = `
"Hello, world!
I am a multiline string."
`;

exports[`printer > print > writes to stdout 1`] = `"Hello, world!"`;

exports[`printer > printTable > writes a table to stdout 1`] = `
"Table Title
Column 1         Column 2        
Row 1, Column 1  Row 1, Column 2 
Row 2, Column 1  Row 2, Column 2 
Table Footer
"
`;

exports[`printer > printTable > writes a table with none borders to stdout 1`] = `
"Table Title
Column 1         Column 2        
Row 1, Column 1  Row 1, Column 2 
Row 2, Column 1  Row 2, Column 2 
Table Footer
"
`;

exports[`printer > printTable > writes a table with thick borders to stdout 1`] = `
"Table Title
╔═════════════════╤═════════════════╗
║ Column 1        │ Column 2        ║
╟─────────────────┼─────────────────╢
║ Row 1, Column 1 │ Row 1, Column 2 ║
╟─────────────────┼─────────────────╢
║ Row 2, Column 1 │ Row 2, Column 2 ║
╚═════════════════╧═════════════════╝
Table Footer
"
`;

exports[`printer > printTable > writes a table with thin borders to stdout 1`] = `
"Table Title
┌─────────────────┬─────────────────┐
│ Column 1        │ Column 2        │
├─────────────────┼─────────────────┤
│ Row 1, Column 1 │ Row 1, Column 2 │
├─────────────────┼─────────────────┤
│ Row 2, Column 1 │ Row 2, Column 2 │
└─────────────────┴─────────────────┘
Table Footer
"
`;

exports[`printer > printTable > writes a table with without a footer to stdout 1`] = `
"Table Title
Column 1         Column 2        
Row 1, Column 1  Row 1, Column 2 
Row 2, Column 1  Row 2, Column 2 
"
`;

exports[`printer > printTable > writes the table as json when GGT_LOG_FORMAT=json 1`] = `
"{\\"level\\":6,\\"name\\":\\"printer\\",\\"msg\\":\\"Table Title\\",\\"fields\\":{\\"headers\\":[\\"Column 1\\",\\"Column 2\\"],\\"rows\\":[[\\"Row 1, Column 1\\",\\"Row 1, Column 2\\"],[\\"Row 2, Column 1\\",\\"Row 2, Column 2\\"]],\\"footer\\":\\"Table Footer\\"}}
"
`;

exports[`printer > println > prints json when GGT_LOG_FORMAT=json 1`] = `
"{\\"level\\":6,\\"name\\":\\"printer\\",\\"msg\\":\\"Hello, world!\\",\\"fields\\":{}}
"
`;

exports[`printer > println > prints like a structured logger when GGT_LOG_LEVEL is set 1`] = `
"12:00:00 PRINT printer:
  Hello, world!
  
"
`;

exports[`printer > println > strips indentation from multiline strings 1`] = `
"Hello, world!
I am a multiline string.
"
`;

exports[`printer > println > supports chalk colors within template literals 1`] = `
"[1mHello, world![22m
[4mI am a multiline string.[24m
"
`;

exports[`printer > println > supports margin top 1`] = `
"
Hello, world!
"
`;

exports[`printer > println > supports tagged template literals 1`] = `
"Hello, world!
I am a multiline string.
"
`;

exports[`printer > println > writes to stdout 1`] = `
"Hello, world!
"
`;
